### DISCUSSION TITLE
General discussion about PHP/Laravel/SQL/etc

### DISCUSSION THREAD

**Discussion by meleu** [RetroAchievements/RAWeb#726] [Ideas]
**This is not an issue nor a feature request!**

Just a thread where we can share some tips and tricks about PHP and thoughts about some parts of RAWeb's code.

**Comment by meleu** [comment]
@luchaos I was taking a look at the [new features of PHP 7](http://php.net/manual/en/migration70.new-features.php) and noticed an interesting operator: [Null coalescing operator](http://php.net/manual/en/migration70.new-features.php#migration70.new-features.null-coalesce-op). `??`

Am I wrong or we could make `seekGET()`, `seekPOST()` and `seekPOSTorGET()` (in the [lib/utils.php](https://github.com/RetroAchievements/RAWeb/blob/master/lib/utils.php)) like this:

```php
function seekGET( $key, $default = NULL )
{
    return $_GET[ $key ] ?? $default;
}

function seekPOST( $key, $default = NULL )
{
    return $_POST[ $key ] ?? $default;
}

function seekPOSTorGET( $key, $default = NULL, $type = NULL )
{
    $retval = $_POST[ $key ] ?? $_GET[ $key ] ?? $default;
    if( isset( $type ) )
        settype( $retval, $type );
    return $retval;
}
```

**Comment by luchaos** [comment]
that does indeed work, very useful as it omits errors as well (like when the check in question is an undefined index of an array)
and makes it shorter than to write multipe ternary operations ($a = $b ? $c : $d)
(the `seek*` functions will be removed in v2 though, much like the rest of the lib folder, as the framework takes care of getting request inputs)

**Comment by meleu** [comment]
@luchaos this part of my code:
```php
    if( isset( $type ) )
        settype( $retval, $type );
```
brought me a question: is there a builtin way to check if an input is a valid type? Or would I need to hardcode the comparison with the known valid types?

**EDIT**: trying to be more clear. I want to avoid situations like `settype( $retval, 'invalid_type' )`.

**Comment by luchaos** [comment]
@meleu with Laravel we will have a Validator at hand with extensive options (see https://laravel.com/docs/5.6/validation)
There will be plenty of such validation rules prepared in the framework branch as a reference.

**Comment by meleu** [comment]
@luchaos can we generate API docs like this?: https://laravel.com/api/5.6/

is it enough to use that javadoc-style comments? this looks really cool and useful! :heart_eyes:

**Comment by luchaos** [comment]
We definitely will have an API documentation, but not like Laravel's internal API as that's for the framework/library API docs which are base on docblocks. The API we're exposing is not for a library but for a web API. 
For web APIs like RA's Web API and App (client) API there are more suitable ways to generate docs, Laravel can do that based on routes, swagger schemas could be an option there as well. We will have those linked on that page for both v1 legacy and the new v2 api, as well as oauth related endpoints: https://stage.retroachievements.org/api
Using docblocks for generating docs would only makes sense for libraries that third parties can integrate - maybe we'll have something like that as well in the future for different frameworks/langs.
